ArduCopter/mode_ship_ops.cpp

```diff
@@ -0,0 +1,754 @@
+        // follow does not have a target
+        gcs().send_text(MAV_SEVERITY_WARNING, "Beacon distance larger than FOLL_DIST_MAX");
+        return false;
+    }
+
+    pos_with_ofs_ned.xy() = curr_pos_neu_cm.xy() + pos_delta_with_ofs_ned_m.xy() * 100.0;
+    pos_with_ofs_ned.z = -curr_pos_neu_cm.z + pos_delta_with_ofs_ned_m.z * 100.0;
+    vel_ned_ms *= 100.0f;
+    
+    float target_heading_deg = 0.0f;
+    g2.follow.get_target_heading_deg(target_heading_deg);
+
+    ship_pos_ned = pos_with_ofs_ned.topostype();
+    ship_vel_ned = vel_ned_ms;
+    ship_accel_ned.zero();
+    ship_heading = radians(target_heading_deg);
+    ship_heading_rate = 0.0;
+    ship_heading_accel = 0.0;
+    ship_available = true;
+
+    offset.zero();
+    offset.xy() = curr_pos_neu_cm.xy() - ship_pos_ned.xy().tofloat();
+
+    // initialise horizontal speed, acceleration
+    pos_control->set_max_speed_accel_xy(wp_nav->get_default_speed_xy(), wp_nav->get_wp_acceleration());
+    pos_control->set_correction_speed_accel_xy(wp_nav->get_default_speed_xy(), wp_nav->get_wp_acceleration());
+
+    // initialize vertical speeds and acceleration
+    pos_control->set_max_speed_accel_z(wp_nav->get_default_speed_down(), wp_nav->get_default_speed_up(), wp_nav->get_accel_z());
+    pos_control->set_correction_speed_accel_z(wp_nav->get_default_speed_down(), wp_nav->get_default_speed_up(), wp_nav->get_accel_z());
+
+    // initialise velocity controller
+    pos_control->init_z_controller();
+    pos_control->init_xy_controller();
+
+    // initialise yaw
+    // auto_yaw.set_mode_to_default(false);
+
+    if (get_alt_hold_state(0.0f) == AltHold_Flying) {
+        // we are flying so we will initialise at the climb to RTL altitude.
+        set_state(SubMode::CLIMB_TO_RTL);
+        ship_takeoff = false;
+    } else {
+        // we are landed so we will initialise in the Final state.
+        set_state(SubMode::LAUNCH_RECOVERY);
+        ship_takeoff = true;
+    }
+
+    return true;
+}
+
+// returns a user-readable string for supplied submode
+const char *ModeShipOperation::state_name(SubMode mode) const
+{
+    switch (mode) {
+    case SubMode::CLIMB_TO_RTL:
+        return "ClimbToRTL";
+    case SubMode::RETURN_TO_PERCH:
+        return "ReturnToPerch";
+    case SubMode::PERCH:
+        return "Perch";
+    case SubMode::OVER_SPOT:
+        return "OverSpot";
+    case SubMode::LAUNCH_RECOVERY:
+        return "LaunchRecovery";
+    case SubMode::PAYLOAD_PLACE:
+        return "PayloadPlace";
+    }
+    return "?";
+}
+
+void ModeShipOperation::set_state(SubMode mode)
+{
+    gcs().send_text(MAV_SEVERITY_WARNING, "ShipOps: %s",
+                    state_name(mode));
+    _state = mode;
+}
+
+void ModeShipOperation::set_keep_out_zone_mode(KeepOutZoneMode keep_out_zone_mode)
+{
+    if (_keep_out_zone_mode != keep_out_zone_mode) {
+        _keep_out_zone_mode = keep_out_zone_mode;
```

Comment:
```suggestion
    if (_keep_out_zone_mode == keep_out_zone_mode) {
        // nothing to do
        return;
    }
    _keep_out_zone_mode = keep_out_zone_mode;
```

(and appropriate reindent)

Reply:
done